app/src/main/java/ch/epfl/polybazaar/category/RootCategory.java

```diff
@@ -0,0 +1,166 @@
+            }
+            catch (Exception e){
+                root = null;
+            }
+        }
+        return root;
+    }
+
+    private RootCategory(Context context) throws IOException, JSONException {
+        nodes = new ArrayList<>();
+            BufferedReader reader = new BufferedReader(new InputStreamReader(context.getAssets().open("categories.json")));
+            String line = reader.readLine();
+            StringBuilder jsonBuilder = new StringBuilder();
+
+            while(line != null){
+                jsonBuilder.append(line).append("\n");
+                line = reader.readLine();
+            }
+            String json = jsonBuilder.toString();
+            JSONObject jsonObject = new JSONObject(new JSONTokener(json));
+            JSONToCategory(jsonObject, this);
+    }
+
+
+    private void JSONToCategory(Object json, Category currentLevelCategory) throws JSONException {
+        if(json instanceof JSONObject){
+            JSONObject jsonObject = (JSONObject)json;
+            Iterator<String> keys = jsonObject.keys();
+            while(keys.hasNext()){
+                String name = keys.next();
+                NodeCategory node = new NodeCategory(name);
+                currentLevelCategory.addSubCategory(node);
+                JSONArray arr = new JSONArray(jsonObject.getJSONArray(name).toString());
+                JSONToCategory(arr, node);
+            }
+        }
+        else if(json instanceof JSONArray){
+            JSONArray jsonArray = (JSONArray)json;
+            for(int i = 0 ; i<jsonArray.length() ; i++){
+                if(!jsonArray.get(i).toString().substring(0,1).equals("{")){
+                    NodeCategory node = new NodeCategory(jsonArray.get(i).toString());
+                    currentLevelCategory.addSubCategory(node);
+                    JSONToCategory(jsonArray.get(i), node);
+                }
+                else{
+                    JSONToCategory(jsonArray.get(i), currentLevelCategory);
+                }
+            }
+        }
+    }
+
+    @Override
+    public void addSubCategory(Category subCategory) {
+        nodes.add(subCategory);
+    }
+
+    @Override
+    public void removeSubCategory(Category subCategory) {
+        nodes.remove(subCategory);
+    }
+
+    @Override
+    public List<Category> subCategories() {
+        return nodes;
+    }
+
+    @Override
+    public boolean hasSubCategories() {
+        return !nodes.isEmpty();
+    }
+
+    @Override
+    public boolean equals(Category other) {
+        return false;
+    }
+
+    @Override
+    public int size(){
+        int size = 0;
+        for(Category category : nodes){
+            size += category.size();
+        }
+        return size;
+    }
+
+    @Override
+    public int maxDepth(){
+        int maxDepth = 0;
+        for(Category category : nodes){
+            maxDepth = max(maxDepth, category.maxDepth());
+        }
+        return maxDepth;
```

Comment:
`return Collections.max(nodes, Comparator.comparing(c -> c.maxDepth()));`

Reply:
Sadly the comparing function requires API version 24 at least which causes an issue with Travis